### DISCUSSION TITLE
we are stuck, method to invalidate token and check if a token is not valid

### DISCUSSION THREAD

**Discussion by Vadilonga** [pocketbase/pocketbase#4181] [Q&A] [answered]
We are trying to implement a way to invalidate all users logged into our application. We are implementing a custom middleware. This is our code:

```
/// <reference path="../pb_data/types.d.ts" />

function myCustomMiddleware(next) {
    return (c) => {
        const auth = c.request().header.get("Authorization");
        console.log('header token', auth)

        try {
            const record = $app.dao().findAuthRecordByToken(
                auth,
                $app.settings().recordAuthToken.secret,
             );
        } catch (err) {
            console.log(err);
            throw new BadRequestError("Invalid token")
        }

        return next(c) // proceed with the request chain
    }
}

routerUse(myCustomMiddleware)
```

Surely we have to manage the routes where the token is not required, for example the login. 
Could there be any problems with this middleware?
Does anyone have a better implementation to achieve the same result?

**We have certainly seen that in the pocketbase panel with this middleware we are no longer able to authenticate.** 
_We could put a check that if you are accessing the route of the pocketbase panel no check is carried out since the panel already takes care of verifying the user._

I dunno, anyone more skilled who wants to give us an implementation idea?

**Comment by ganigeorgiev** [comment]
The above middleware doesn't invalidate tokens - it just throw an error if no valid record auth token was sent with the request and if you register it globally as in the above it will always prevent admin authentication.

If you want to invalidate all issued record auth tokens you can set a new value for the `$app.settings().recordAuthToken.secret`.

```js
const settings = $app.settings()
settings.recordAuthToken.secret = $security.randomString(50)
$app.dao().saveSettings(settings)
```

**Reply by ganigeorgiev** [reply]
_It is not clear from your description but if you are looking for a middleware that will allow access to a specific action **only** for an authenticated auth record you can use the `$apis.requireRecordAuth("users")` middleware (see [Builtin middlewares](https://pocketbase.io/docs/js-routing/#builtin-middlewares))._

**Reply by Vadilonga** [reply]
I agree with what you told me. We invalidate the tokens from the interface, we just want to check that the token that comes to us is actually valid, if it is not we return a 401 and have the user log in again from the UI. 
The problem is that we are using the routerUse function which applies the middleware globally on all routes. 
So if we wanted to apply it to all db collections except api such as login, registration or the pocketbase admin panel we have to write a custom middleware or is there anything built-in solution?

**Reply by ganigeorgiev** [reply]
I'm failing to understand what you are trying to do. Please check the documentation and the available builtin middlewares at https://pocketbase.io/docs/js-routing/#middlewares.

If that doesn't help, please try to elaborate with a simple example what is the desired behavior and the expected execution flow.

**Reply by ganigeorgiev** [reply]
Also if you want to apply auth restrictions to the existing collection web APIs, prefer using the API rules -https://pocketbase.io/docs/api-rules-and-filters/.

**Reply by Vadilonga** [reply]
Sorry, I'll try to explain what I wanted with a user story:
**goal**: 
- "As an administrator I want to be able to invalidate my users' tokens."

**details**:
- "When I invalidate a token the involved user who is using the mobile application must be redirected to the login page."


**current situation without rules**:
- "If I invalidate the tokens, users can continue browsing the application."

**first test - apply restrictions on the api**:
- "Users can navigate the application, pocketbase, upon request for resources, if the user token is not valid returns an empty list"
feedback: "I cannot apply a check on 'response body is empty' to redirect the user to the login page"

**second test - apply the middleware to validate token**:
- "The user can only navigate when he has a valid token. However, when I invalidate the token you no longer have the possibility to access the login page"

I hope I have provided a comprehensive overview. I apologize if my initial question was unclear.

**Reply by ganigeorgiev** [reply, marked as answer]
I think you are using _"invalidate the token"_ in the above context incorrectly but I'll not pursuit the matter further.

If after invalidating the user tokens from the Admin UI (or by changing the settings secret as shown above) you want to force your users to logout from your client-side app, then you can send [`pb.collection("users").authRefresh()` call](https://pocketbase.io/docs/api-records/#auth-refresh) on each page/app load.

If you want to always throw an error whenever the user provided an invalid or expired token, then you can use the following middleware:

```js
// throw an error if the request has an Authorization:TOKEN header but
// it is invalid or expired (aka. there is no auth model found)
routerUse((next) => {
    return (c) => {
        if (
            // has token header
            c.request().header.get("Authorization") &&
            // there is no valid auth model record associated with the token
            !c.get("admin") &&
            !c.get("authRecord")
        ) {
            throw new UnauthorizedError("Invalid or expired token.")
        }

        return next(c);
    }
})
```

**Reply by Vadilonga** [reply]
Ok, ok then I have to check in the frontend application if the user's current token is valid or not.

I'm using invalidate the token because in the pocketbase controller the action that we use to invalidate the token is this one:
![image](https://github.com/pocketbase/pocketbase/assets/51320388/829e90fb-dfda-47ba-a908-e13ebe87cd22)

I know that this way I invalidate the token of all logged in users.

So, the best practice is not to use middleware but to check the validity of the token every time a route is made in the mobile application, right?

Thanks for the support, pocketbase is awesome, we really appreciate the work you do!

**Reply by ganigeorgiev** [reply]
I think there is a language barrier that prevents me to understand what you asking and what you are trying to do (I'm not a native speaker).

> So, the best practice is not to use middleware but to check the validity of the token every time a route is made in the mobile application, right?

No, there is no best practice here. If you want to always throw an error whenever a request with invalid or expired token is sent use the above provided middleware.

**Comment by Vadilonga** [comment]
Yes, I'm sorry, I'm not a native speaker either and I think my english is not the best :(

Thank you for your support, we will erect a statue in your honor!

**Reply by jonshipman** [reply]
Put a user login form on your 403 error page. When you invalidate all the tokens, the next request the user will make will result in a 403 error -> showing the login form. Personal preference as it keeps the HTTP status codes intact (if you redirect, you change the 403 to a 302).

You can use a heartbeat and ping the server every x seconds if you want a user to be logged out without an action on their part. Or create a collection that has a single record with a count column and a pb_hook that increments when the token is invalidated. Then in your client, you can subscribe to that collection and it will ping when the count is updated.

**Reply by Vadilonga** [reply]
We have found a solution, in our opinion it could be the best choice:

- We have created a custom middleware that triggers only on the recovering data endpoints . If the token is not verified we return a 401.
- In our app we have created a custom fetch that allows you to intercept the 401 error with the angular interceptor.
- When we intercept a 401 we send the user back to the login page.

**Done**

*If anyone has similar needs I will be happy to share our implementations!
